File: oneapi-rs-sys/src/platform.cpp
#include "oneapi-rs-sys/include/platform.hpp"

namespace sycl_shims {
std::unique_ptr<std::vector<Platform>> Platform::get_platforms() {
  std::vector<Platform> platforms;
  for (auto &&platform : sycl::platform::get_platforms())
    platforms.push_back(platform);

  return std::make_unique<std::vector<Platform>>(platforms);
}

rust::String Platform::get_version() const {
  return this->inner.get_info<sycl::info::platform::version>();
}

rust::String Platform::get_name() const {
  return this->inner.get_info<sycl::info::platform::name>();
}

rust::String Platform::get_vendor() const {
  return this->inner.get_info<sycl::info::platform::vendor>();
}
} // namespace sycl_shims

File: oneapi-rs-sys/src/platform.rs
#[cxx::bridge(namespace = "sycl_shims")]
pub mod ffi {
    // This is a workaround - cxx currently doesn't support passing
    // around C++ vectors of pointers directly
    // https://github.com/dtolnay/cxx/issues/774#issuecomment-808674945
    struct PlatformPtr {
        // This has to be a SharedPtr because a shared struct needs to be
        // copy-constructible
        ptr: SharedPtr<Platform>
    }

    unsafe extern "C++" {
        include!("oneapi-rs-sys/include/platform.hpp");
 
        type Platform;

        #[Self = "Platform"]
        fn get_platforms() -> UniquePtr<CxxVector<PlatformPtr>>;

        fn get_version(&self) -> String;
        fn get_name(&self) -> String;
        fn get_vendor(&self) -> String;
    }
}
